File: code/backend/Odin/src/main/kotlin/pt/isel/odin/service/interfaces/CategoryService.kt
package pt.isel.odin.service.interfaces

import pt.isel.odin.model.Category

/**
 * Service for Categories.
 */
interface CategoryService {

    /**
     * Gets a Category by its id.
     *
     * @param id the curricular unit id.
     */
    fun getById(id: Long): Category

    /**
     * Gets all Categories.
     */
    fun getAll(): List<Category>

    /**
     * Saves or Updates a Category.
     *
     * @param categoryRequest the curricular unit to save.
     */
    fun save(categoryRequest: Category): Category

    /**
     * Deletes a Category by its id.
     *
     * @param id the Category id.
     */
    fun delete(id: Long)
}

File: code/frontend/Tyr/src/services/CategoryService.ts
import {makeApiRequest} from '../axios/apiRequest';
import {Category} from "../model/Category";

/**
 * Service to manage categories
 */
export class CategoryService {

    static basePath = '/categories';

    static async getById(id: number): Promise<Category> {
        return makeApiRequest('get', `${(CategoryService.basePath)}/${id}`);
    }

    static async getAll(): Promise<Category[]> {
        return makeApiRequest('get', CategoryService.basePath);
    }

    static async save(curricularUnitRequest: Category): Promise<Category> {
        return makeApiRequest('post', `${(CategoryService.basePath)}/save`, curricularUnitRequest);
    }

    static async update(curricularUnitRequest: Category): Promise<Category> {
        return makeApiRequest('put', `${(CategoryService.basePath)}/update`, curricularUnitRequest);
    }

    static async delete(id: number): Promise<void> {
        return makeApiRequest('delete', `${(CategoryService.basePath)}/${id}`);
    }
}